internal/promotion/runner/builtin/yaml_merger.go

```diff
@@ -0,0 +1,152 @@
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+	"github.com/xeipuuv/gojsonschema"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/yaml"
+	"github.com/akuity/kargo/pkg/promotion"
+	"github.com/akuity/kargo/pkg/x/promotion/runner/builtin"
+)
+
+// yamlMerger is an implementation of the promotion.StepRunner interface that
+// updates do a merge of the multiple YAML files.
+type yamlMerger struct {
+	schemaLoader gojsonschema.JSONLoader
+}
+
+// newYAMLMerger returns an implementation of the promotion.StepRunner interface
+// that updates the values of specified keys in a YAML file.
+func newYAMLMerger() promotion.StepRunner {
+	r := &yamlMerger{}
+	r.schemaLoader = getConfigSchemaLoader(r.Name())
+	return r
+}
+
+// Name implements the promotion.StepRunner interface.
+func (y *yamlMerger) Name() string {
+	return "yaml-merge"
+}
+
+// Run implements the promotion.StepRunner interface.
+func (y *yamlMerger) Run(
+	ctx context.Context,
+	stepCtx *promotion.StepContext,
+) (promotion.StepResult, error) {
+	failure := promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored}
+
+	if err := y.validate(stepCtx.Config); err != nil {
+		return failure, err
+	}
+
+	// Convert the configuration into a typed struct
+	cfg, err := promotion.ConfigToStruct[builtin.YAMLMergeConfig](stepCtx.Config)
+	if err != nil {
+		return failure, fmt.Errorf("could not convert config into %s config: %w", y.Name(), err)
+	}
+
+	return y.run(ctx, stepCtx, cfg)
+}
+
+// validate validates yamlMerger configuration against a JSON schema.
+func (y *yamlMerger) validate(cfg promotion.Config) error {
+	return validate(y.schemaLoader, gojsonschema.NewGoLoader(cfg), y.Name())
+}
+
+func (y *yamlMerger) run(
+	_ context.Context,
+	stepCtx *promotion.StepContext,
+	cfg builtin.YAMLMergeConfig,
+) (promotion.StepResult, error) {
+
+	result := promotion.StepResult{Status: kargoapi.PromotionStepStatusSucceeded}
+	failure := promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored}
+
+	mergedFiles := []string{} // keep track of files actually merged
+
+	// Secure join the paths to prevent path traversal attacks.
+	yamlData := []string{}
+	for _, path := range cfg.InPaths {
+		inPath, err := securejoin.SecureJoin(stepCtx.WorkDir, path)
+		if err != nil {
+			return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+				fmt.Errorf("could not secure join inPath %q: %w", path, err)
+		}
+
+		inBytes, err := os.ReadFile(inPath)
+		if err != nil {
+			// we skip if file does not exist
+			if cfg.IgnoreMissingFiles && os.IsNotExist(err) {
+				continue
+			}
+			return failure, fmt.Errorf(
+				"error reading file %q: %w",
+				inPath,
+				err,
+			)
+		}
+
+		// we skip if file is empty
+		if len(inBytes) == 0 {
+			continue
+		}
+
+		mergedFiles = append(mergedFiles, path)
+		yamlData = append(yamlData, string(inBytes))
+	}
+
+	// merge YAML files
```

Comment:
Nit: You don't have files in-hand here. You're merging data that's in-memory.

```suggestion
	// Merge
```